README.md
[![Stand With Ukraine](https://raw.githubusercontent.com/vshymanskyy/StandWithUkraine/main/banner2-direct.svg)](https://vshymanskyy.github.io/StandWithUkraine/)

<div align="center">
  <h1>:point_right:<a href="https://github.com/nestjsx/crud/issues/784">You can help NestJs CRUD
  </a></h1>
</div>

<div align="center">
  <h1>CRUD</h1>
</div>
<div align="center">
  <strong>for RESTful APIs built with NestJs</strong>
</div>

<br />

<div align="center">
  <a href="https://github.com/vshymanskyy/StandWithUkraine/blob/main/docs/README.md">
    <img src="https://raw.githubusercontent.com/vshymanskyy/StandWithUkraine/main/badges/StandWithUkraine.svg" alt="StandWithUkraine" />
  </a>
  <a href="https://travis-ci.org/nestjsx/crud">
    <img src="https://github.com/nestjsx/crud/workflows/Tests/badge.svg" alt="Build" />
  </a>
  <a href="https://coveralls.io/github/nestjsx/crud?branch=master">
    <img src="https://coveralls.io/repos/github/nestjsx/crud/badge.svg" alt="Coverage" />
  </a>
  <a href="https://github.com/nestjsx/crud/blob/master/LICENSE">
    <img src="https://img.shields.io/github/license/nestjsx/crud.svg" alt="License" />
  </a>
  <a href="https://www.npmjs.com/package/@fuhidev/crud">
    <img src="https://img.shields.io/npm/v/@fuhidev/crud.svg" alt="npm version" />
  </a>
  <a href="https://www.npmjs.com/org/nestjsx">
    <img src="https://img.shields.io/npm/dm/@fuhidev/crud.svg" alt="npm downloads" />
  </a>
  <a href="http://makeapullrequest.com">
    <img src="https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square" alt="PRs welcome" />
  </a>
  <a href="https://github.com/marmelab/awesome-rest#nodejs">
    <img src="https://raw.githubusercontent.com/nestjsx/crud/master/img/awesome-rest.svg?sanitize=true" alt="Awesome REST" />
  </a>
  <a href="#individuals" alt="Sponsors on Open Collective">
    <img src="https://opencollective.com/nestjsx/backers/badge.svg" />
  </a>
  <a href="#organizations" alt="Sponsors on Open Collective">
    <img src="https://opencollective.com/nestjsx/sponsors/badge.svg" />
  </a> 
</div>

<div align="center">
  <sub>Built with :purple_heart: by
  <a href="https://twitter.com/MichaelYali">@MichaelYali</a> and
  <a href="https://github.com/nestjsx/crud/graphs/contributors">
    Contributors
  </a>
  <div align="center">
    :star2: :eyes: :zap: :boom:
  </div>
</div>

<br />

We believe that everyone who's working with NestJs and building some RESTful services and especially some CRUD functionality will find `@fuhidev/crud` microframework very useful.

## Features

<img align="right" src="img/crud-usage2.png" alt="CRUD usage" />

- :electric_plug: Super easy to install and start using the full-featured controllers and services :point_right:

- :octopus: DB and service agnostic extendable CRUD controllers

- :mag_right: Reach query parsing with filtering, pagination, sorting, relations, nested relations, cache, etc.

- :telescope: Framework agnostic package with query builder for a frontend usage

- :space_invader: Query, path params and DTOs validation included

- :clapper: Overriding controller methods with ease

- :wrench: Tiny config (including globally)

- :gift: Additional helper decorators

- :pencil2: Swagger documentation

## Packages

- [**@fuhidev/crud**](https://www.npmjs.com/package/@fuhidev/crud) - core package which provides `@Crud()` decorator for endpoints generation, global configuration, validation, helper decorators ([docs](https://github.com/nestjsx/crud/wiki/Controllers#description))
- [**@fuhidev/crud-request**](https://www.npmjs.com/package/@fuhidev/crud-request) - request builder/parser package which provides `RequestQueryBuilder` class for a frontend usage and `RequestQueryParser` that is being used internally for handling and validating query/path params on a backend side ([docs](https://github.com/nestjsx/crud/wiki/Requests#frontend-usage))
- [**@fuhidev/crud-typeorm**](https://www.npmjs.com/package/@fuhidev/crud-typeorm) - TypeORM package which provides base `TypeOrmCrudService` with methods for CRUD database operations ([docs](https://github.com/nestjsx/crud/wiki/ServiceTypeorm))

## Documentation

- :dart: [General Information](https://github.com/nestjsx/crud/wiki#why)
- :video_game: [CRUD Controllers](https://github.com/nestjsx/crud/wiki/Controllers#description)
- :horse_racing: [CRUD ORM Services](https://github.com/nestjsx/crud/wiki/Services#description)
- :trumpet: [Handling Requests](https://github.com/nestjsx/crud/wiki/Requests#description)

## Support

Any support is welcome. At least you can give us a star :star:

## Contributors

### Code Contributors

This project exists thanks to all the people who contribute. [[Contribute](CODE_OF_CONDUCT.md)].
<a href="https://github.com/nestjsx/crud/graphs/contributors"><img src="https://opencollective.com/nestjsx/contributors.svg?width=890&button=false" /></a>

### Financial Contributors

Become a financial contributor and help us sustain our community. [[Contribute](https://opencollective.com/nestjsx#backer)]

#### Individuals

<a href="https://opencollective.com/nestjsx#backers" target="_blank"><img src="https://opencollective.com/nestjsx/backers.svg?width=890&button=false"></a>

#### Organizations

Support this project with your organization. Your logo will show up here with a link to your website. [[Contribute](https://opencollective.com/nestjsx#sponsor)]

<a href="https://opencollective.com/nestjsx/sponsor/0/website" target="_blank"><img src="https://opencollective.com/nestjsx/sponsor/0/avatar.svg"></a>
<a href="https://opencollective.com/nestjsx/sponsor/1/website" target="_blank"><img src="https://opencollective.com/nestjsx/sponsor/1/avatar.svg"></a>
<a href="https://opencollective.com/nestjsx/sponsor/2/website" target="_blank"><img src="https://opencollective.com/nestjsx/sponsor/2/avatar.svg"></a>
<a href="https://opencollective.com/nestjsx/sponsor/3/website" target="_blank"><img src="https://opencollective.com/nestjsx/sponsor/3/avatar.svg"></a>
<a href="https://opencollective.com/nestjsx/sponsor/4/website" target="_blank"><img src="https://opencollective.com/nestjsx/sponsor/4/avatar.svg"></a>
<a href="https://opencollective.com/nestjsx/sponsor/5/website" target="_blank"><img src="https://opencollective.com/nestjsx/sponsor/5/avatar.svg"></a>
<a href="https://opencollective.com/nestjsx/sponsor/6/website" target="_blank"><img src="https://opencollective.com/nestjsx/sponsor/6/avatar.svg"></a>
<a href="https://opencollective.com/nestjsx/sponsor/7/website" target="_blank"><img src="https://opencollective.com/nestjsx/sponsor/7/avatar.svg"></a>
<a href="https://opencollective.com/nestjsx/sponsor/8/website" target="_blank"><img src="https://opencollective.com/nestjsx/sponsor/8/avatar.svg"></a>
<a href="https://opencollective.com/nestjsx/sponsor/9/website" target="_blank"><img src="https://opencollective.com/nestjsx/sponsor/9/avatar.svg"></a>

## License

[MIT](LICENSE)


CHANGELOG.md
## [4.6.2] - 2020-05-14

### Bug Fixes

- **typeorm** - fixed selected fields on joins [#510](https://github.com/nestjsx/crud/issues/510)). Kudos to @jbrousseau for finding this bug

## [4.6.1] - 2020-05-08

### Bug Fixes

- **typeorm** - fixed query generation when a column display name differs from its name in db [#401](https://github.com/nestjsx/crud/issues/401)). Kudos to @farhad2161 for finding this bug

## [4.6.0] - 2020-05-07

### Features

- **crud**/**typeorm** - added `select` (boolean) to `join` options which allows to join relation but not select it ([#218](https://github.com/nestjsx/crud/issues/218))

### Bug Fixes

- **typeorm** - fixed column identifier for MySQL ([#401](https://github.com/nestjsx/crud/issues/401))
- **typeorm** - fixed nested relations aliases, filtering, sorting ([#419](https://github.com/nestjsx/crud/issues/419), [#450](https://github.com/nestjsx/crud/issues/450), [#267](https://github.com/nestjsx/crud/issues/267), [#385](https://github.com/nestjsx/crud/issues/385))

## [4.5.0] - 2020-05-01

### Improvements

- **crud** - added enum support for params Swagger. Kudos to @tbrannam

### Bug Fixes

- **crud** - fixed auth property definition. Kudos to @lafeuil
- **typeorm** - fixed request generation with aliases ([#321](https://github.com/nestjsx/crud/issues/321), [#401](https://github.com/nestjsx/crud/issues/401)). Kudos to @joennlae

## [4.4.5] - 2020-04-18

## Deps

- **crud** fixted imports

## [4.4.4] - 2020-04-18

### Deps

- **dev** fixted lerna

## [4.4.3] - 2020-04-18

### Bug Fixes

- **crud** fixed returning `pageCount` in some cases ([#465](https://github.com/nestjsx/crud/pull/465))
- **typeorm** fixed critical bug with possible SQL injections when using query `?sort=` (big kudos to João Maurício)
- **typeorm** fixed filter conditions for LIKE/iLIKE operators ([#395](https://github.com/nestjsx/crud/pull/395))

## [4.4.2] - 2020-03-17

### Bug Fixes

- **crud** fixed custom routes params caused by NestJs v7 breaking changes ([#443](https://github.com/nestjsx/crud/issues/443))

## [4.4.1] - 2019-12-28

### Bug Fixes

- **crud** fixed `CrudRequestInterceptor` validation status code from 500 to 400 ([#374](https://github.com/nestjsx/crud/issues/374), [#247](https://github.com/nestjsx/crud/issues/247))

## [4.4.0] - 2019-12-27

### Features

- **crud** added `serialize` to the global options

## [4.3.0] - 2019-12-21

### Features

- **crud** added `dto` to the `CrudOptions` ([#132](https://github.com/nestjsx/crud/issues/132))
- **crud** added `serialize` to the `CrudOptions`
- **crud** added `search` query param and a new search condition api
- **crud** added new condition operators: `$eqL`, `$neL`, `$startsL`, `$endsL`, `$contL`, `$exclL`, `$inL`, `$notinL` for case insensitive queries ([#77](https://github.com/nestjsx/crud/issues/77))
- **crud** added `@crudAuth()` class decorator for authorized requests

### Improvements

- **crud** `CrudRequestInterceptor` can be used for both crud and non-crud controllers or for custom routes within crud controller
- **crud** support `@nestjs/swagger` major versions: v3 and v4 ([#340](https://github.com/nestjsx/crud/issues/340))
- **crud** added `returnShallow` option to the `CrudOptions.routes` `createOneBase`, `updateOneBase`, `replaceOneBase` methods ([#158](https://github.com/nestjsx/crud/issues/158))
- **crud** added `alias` to the `CrudOptions.join` ([#350](https://github.com/nestjsx/crud/issues/55))
- **crud** added `alwaysPaginate` to the `CrudOptions.query`, can be used globally as well ([#213](https://github.com/nestjsx/crud/issues/213))
- **crud** `CrudOptions.query.filter` can be a function that returns transformed `search` object
- **crud** added `disabled` for an objects withing `CrudOptions.params`
- **request** query builder: now uses [qs](https://www.npmjs.com/package/qs) package
- **request** query builder: `filter` and `or` methods can accept array of filter objects
- **typeorm** changed visibility of all methods ([#226](https://github.com/nestjsx/crud/issues/226))
- **typeorm** use `ILIKE` for PostgreSQL ([#212](https://github.com/nestjsx/crud/issues/212))

### Bug Fixes

- **crud** swagger: fixed response models ([#350](https://github.com/nestjsx/crud/issues/350))
- **crud** swagger: fixed query params ([#196](https://github.com/nestjsx/crud/issues/196))
- **crud** swagger: fixed renamed params ([#283](https://github.com/nestjsx/crud/issues/283))
- **crud** swagger: fixed swagger method decoration on overridden methods
- **crud** query parser: fixed parsing integers when it's a big int
- **typeorm** fixed load embedded entities ([#138](https://github.com/nestjsx/crud/issues/138))
- **typeorm** fixed left join issues ([#31](https://github.com/nestjsx/crud/issues/31), [#98](https://github.com/nestjsx/crud/issues/98))
- **typeorm** fixed composite key joins ([#238](https://github.com/nestjsx/crud/issues/238))
- **typeorm** fixed entity events ([#51](https://github.com/nestjsx/crud/issues/51))
- **typeorm** all methods return entity instances ([#259](https://github.com/nestjsx/crud/issues/259))

## [4.2.0] - 2019-07-26

### Features

- **crud** added support for older versions of `UUID` ([#186])

### Bug Fixes

- **crud** fixed `BulkDto` swagger description ([#159])
- **crud** fixed `CrudRequestInterceptor` request parsing
- **requests** added `@fuhidev/util` as a dependency ([#184])
- **requests** fixed condition operators mapping ([#148])
- **requests** fixed ISO date string validation ([#161])
- **typeorm** fixed filtering and sorting by nested fields ([#105])
- **typeorm** fixed `too many nested levels` exception ([#87])
- **typeorm** fixed pagination `pageCount` ([#179])

### Deps

- **dev** updated deps

## [4.1.0] - 2019-06-27

### Features

- **crud** added `PUT` request handling ([#107])
- **requests** added creating request builder with params ([#131])
- **requests** improved query params naming parsing ([#101])

### Bug Fixes


integration/crud-typeorm/seeds.ts
import { ClassType } from '@fuhidev/util';
import { plainToClass } from 'class-transformer';
import { MigrationInterface, QueryRunner, Repository } from 'typeorm';
import { Company } from './companies';
import { Note } from './notes';
import { Project, UserProject } from './projects';
import { Name, User } from './users';
import { License, UserLicense } from './users-licenses';
import { UserProfile } from './users-profiles';

export class Seeds1544303473346 implements MigrationInterface {
  private save<T>(repo: Repository<T>, data: Partial<T>[]): Promise<T[]> {
    return repo.save(
      data.map((partial: Partial<T>) =>
        plainToClass<any, any>(repo.target as ClassType<T>, partial, {
          ignoreDecorators: true,
        }),
      ),
    );
  }

  public async up(queryRunner: QueryRunner): Promise<any> {
    const { connection } = queryRunner;

    const companiesRepo = connection.getRepository(Company);
    const projectsRepo = connection.getRepository(Project);
    const usersProfilesRepo = connection.getRepository(UserProfile);
    const usersRepo = connection.getRepository(User);
    const licensesRepo = connection.getRepository(License);
    const usersLincesesRepo = connection.getRepository(UserLicense);
    const usersProjectsRepo = connection.getRepository(UserProject);
    const notesRepo = connection.getRepository(Note);

    // companies
    await this.save(companiesRepo, [
      { name: 'Name1', domain: 'Domain1' },
      { name: 'Name2', domain: 'Domain2' },
      { name: 'Name3', domain: 'Domain3' },
      { name: 'Name4', domain: 'Domain4' },
      { name: 'Name5', domain: 'Domain5' },
      { name: 'Name6', domain: 'Domain6' },
      { name: 'Name7', domain: 'Domain7' },
      { name: 'Name8', domain: 'Domain8' },
      { name: 'Name9', domain: 'Domain9', deletedAt: new Date() },
      { name: 'Name10', domain: 'Domain10' },
    ]);

    // projects
    await this.save(projectsRepo, [
      { name: 'Project1', description: 'description1', isActive: true, companyId: 1 },
      { name: 'Project2', description: 'description2', isActive: true, companyId: 1 },
      { name: 'Project3', description: 'description3', isActive: true, companyId: 2 },
      { name: 'Project4', description: 'description4', isActive: true, companyId: 2 },
      { name: 'Project5', description: 'description5', isActive: true, companyId: 3 },
      { name: 'Project6', description: 'description6', isActive: true, companyId: 3 },
      { name: 'Project7', description: 'description7', isActive: true, companyId: 4 },
      { name: 'Project8', description: 'description8', isActive: true, companyId: 4 },
      { name: 'Project9', description: 'description9', isActive: true, companyId: 5 },
      { name: 'Project10', description: 'description10', isActive: true, companyId: 5 },
      { name: 'Project11', description: 'description11', isActive: false, companyId: 6 },
      { name: 'Project12', description: 'description12', isActive: false, companyId: 6 },
      { name: 'Project13', description: 'description13', isActive: false, companyId: 7 },
      { name: 'Project14', description: 'description14', isActive: false, companyId: 7 },
      { name: 'Project15', description: 'description15', isActive: false, companyId: 8 },
      { name: 'Project16', description: 'description16', isActive: false, companyId: 8 },
      { name: 'Project17', description: 'description17', isActive: false, companyId: 9 },
      { name: 'Project18', description: 'description18', isActive: false, companyId: 9 },
      { name: 'Project19', description: 'description19', isActive: false, companyId: 10 },
      { name: 'Project20', description: 'description20', isActive: false, companyId: 10 },
    ]);

    // user-profiles
    await this.save(usersProfilesRepo, [
      { name: 'User1' },
      { name: 'User2' },
      { name: 'User3' },
      { name: 'User4' },
      { name: 'User5' },
      { name: 'User6' },
      { name: 'User7' },
      { name: 'User8' },
      { name: 'User9' },
      { name: 'User1' },
      { name: 'User1' },
      { name: 'User1' },
      { name: 'User1' },
      { name: 'User1' },
      { name: 'User1' },
      { name: 'User1' },
      { name: 'User1' },
      { name: 'User1' },
      { name: 'User1' },
      { name: 'User2' },
    ]);

    // users
    const name: Name = { first: null, last: null };
    const name1: Name = { first: 'firstname1', last: 'lastname1' };
    await this.save(usersRepo, [
      { email: '1@email.com', isActive: true, companyId: 1, profileId: 1, name: name1 },
      { email: '2@email.com', isActive: true, companyId: 1, profileId: 2, name },
      { email: '3@email.com', isActive: true, companyId: 1, profileId: 3, name },
      { email: '4@email.com', isActive: true, companyId: 1, profileId: 4, name },
      { email: '5@email.com', isActive: true, companyId: 1, profileId: 5, name },
      { email: '6@email.com', isActive: true, companyId: 1, profileId: 6, name },
      { email: '7@email.com', isActive: false, companyId: 1, profileId: 7, name },
      { email: '8@email.com', isActive: false, companyId: 1, profileId: 8, name },
      { email: '9@email.com', isActive: false, companyId: 1, profileId: 9, name },
      { email: '10@email.com', isActive: true, companyId: 1, profileId: 10, name },
      { email: '11@email.com', isActive: true, companyId: 2, profileId: 11, name },
      { email: '12@email.com', isActive: true, companyId: 2, profileId: 12, name },
      { email: '13@email.com', isActive: true, companyId: 2, profileId: 13, name },
      { email: '14@email.com', isActive: true, companyId: 2, profileId: 14, name },
      { email: '15@email.com', isActive: true, companyId: 2, profileId: 15, name },
      { email: '16@email.com', isActive: true, companyId: 2, profileId: 16, name },
      { email: '17@email.com', isActive: false, companyId: 2, profileId: 17, name },
      { email: '18@email.com', isActive: false, companyId: 2, profileId: 18, name },
      { email: '19@email.com', isActive: false, companyId: 2, profileId: 19, name },
      { email: '20@email.com', isActive: false, companyId: 2, profileId: 20, name },
      { email: '21@email.com', isActive: false, companyId: 2, profileId: null, name },
    ]);

    // licenses
    await this.save(licensesRepo, [
      { name: 'License1' },
      { name: 'License2' },
      { name: 'License3' },
      { name: 'License4' },
      { name: 'License5' },
    ]);

    // user-licenses
    await this.save(usersLincesesRepo, [
      { userId: 1, licenseId: 1, yearsActive: 3 },
      { userId: 1, licenseId: 2, yearsActive: 5 },
      { userId: 1, licenseId: 4, yearsActive: 7 },
      { userId: 2, licenseId: 5, yearsActive: 1 },
    ]);

    // user-projects
    await this.save(usersProjectsRepo, [
      { projectId: 1, userId: 1, review: 'User project 1 1' },
      { projectId: 1, userId: 2, review: 'User project 1 2' },
      { projectId: 2, userId: 2, review: 'User project 2 2' },
      { projectId: 3, userId: 3, review: 'User project 3 3' },
    ]);

    // notes
    await this.save(notesRepo, [
      { revisionId: 1 },
      { revisionId: 1 },
      { revisionId: 2 },
      { revisionId: 2 },
      { revisionId: 3 },
      { revisionId: 3 },
    ]);
  }

  public async down(queryRunner: QueryRunner): Promise<any> {}
}
